### PR TITLE
Show diff in coverage for change request builds

### PR DESCRIPTION
**Pull request by tempora-mutantur**
It would be very convenient to see the code coverage diff right in Jenkins. If our
build is going for a multibranch change request job  - we take the
target branch commit that is used for our change request build. Then, we search through the
last fixed amount of builds the build that was built for this target branch commit. If
we find it - we check if it has line code coverage information. In case it has it -  we calculate diff between the current code coverage and this target branch build coverage.

Examples:
Increased code coverage:
image

Decreased code coverage:
image

No changes:
image

Thank you.

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java [hunk 1]
```diff
@@ -143,6 +152,65 @@ private void convertResultToAction(CoverageResult coverageReport, HealthReport h
         }
     }
 
+    private void setDiffInCoverageForChangeRequest(CoverageResult coverageReport) {
+        // Use limited number of builds to don't search too long
+        final int numberOfBuildsFromTargetBranch = 50;
+        // Calculate diff only for line coverage
+        Ratio changeRequestLinesCoverage = coverageReport.getCoverage(CoverageElement.LINE);
+        if (changeRequestLinesCoverage != null) {
```
Comment (cizezsy) [inline src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: Could you reduce the nesting level? There are too many nested if statement and hard to recognize.

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java [hunk 2]
```diff
@@ -97,6 +105,7 @@ public void performCoverageReport(List<CoverageReportAdapter> reportAdapters, Li
         HealthReport healthReport = processThresholds(results, globalThresholds);
 
 
+        setDiffInCoverageForChangeRequest(coverageReport);
```
Comment (cizezsy) [inline src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: Should we make it a configurable option?

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java (hunk 1 above)
Comment (tempora-mutantur) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: Agree, will try

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java (hunk 2 above)
Comment (tempora-mutantur) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: Yes, I think it would be better to make it configurable. Will change it.

Comment (cizezsy) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: @tandibar I think we should config it in Publisher level :)

Comment (tempora-mutantur) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: @cizezsy Please, check

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java (hunk 1 above)
Comment (tempora-mutantur) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: @cizezsy Please, check

Comment (cizezsy) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: @tempora-mutantur Maybe we can use some codes like this
```
if (changeRequestLinesCoverage == null) {
    return
}

Job job = run.getParent();
parent = job.getParent();

if (!(parent instanceof MultiBranchProject)) {
    return
}

......
```

Your current approach will pass through all if statement even if the condition must be false.

File: src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java [hunk 3]
```diff
@@ -165,6 +168,15 @@ public void setTag(String tag) {
         this.tag = tag;
     }
 
+    public boolean getCalculateDiffForChangeRequests() {
+        return calculateDiffForChangeRequests;
+    }
+
+    @DataBoundSetter
+    public void setCalculateDiffForChangeRequests(boolean calculateDiffForChangeRequests) {
```
Comment (cizezsy) [inline src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java]: We need to add a checkbox in CoveragePublisher/config.jelly to config it.

File: src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java (hunk 1 above)
Comment (tempora-mutantur) [reply src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java]: @cizezsy 
Thank you. Please, take a look, I've made such change

**Comment by TinyZzh** [comment]
is this just diff numeric with two branch ?
 
why not we analysis the total detail diff coverage with each file?  
for example: usually a PR includes some git commits.  and we want to known, how much code coverage increase or decrease by this PR. just with this commits.    

@cizezsy @tempora-mutantur

**Comment by tempora-mutantur** [comment]
> why not we analysis the total detail diff coverage with each file?

Sorry, I'm not sure what you mean. This calculates diff between your PR **build** and the target branch **build** that was built on a commit that your PR is based on. This final number, the diff, is the total line coverage diff, one of the most important things that tools like codecov and coveralls do. So you can understand if your PR reduces/increases code coverage. If you build every commit of your PR you'll have the statistics for every commit.

Diff for each file can take a lot of time, I mean, machine time, if I understand this correctly. We may add this functionality further, but this should be optimised really hard. Tools like codecov and coveralls do this asynchronously.

**Comment by TinyZzh** [comment]
sorry for my poor English.  
my company have a large and old java project.  we want calculate PR's changed code's coverage without the previous code. a little like codecov's relative coverage feature.  

found the codecov's docs.  use git diff command to calculates relative code coverage. docs link:
 [CodeCov's feature comparing-commits](https://docs.codecov.io/docs/comparing-commits) .

**Comment by tempora-mutantur** [comment]
@TinyZzh Sorry, I don't understand what you mean by "without the previous code". Actually, you can think about this that this is a diff between two commits, one from the target branch and the other from is the PR head.

**Comment by tempora-mutantur** [comment]
@cizezsy Hi. Are there any other problems with this PR? Maybe we can think of them together?

File: src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java [hunk 4]
```diff
@@ -49,6 +49,8 @@
 
     private String tag;
 
+    private boolean calculateDiffForChangeRequests = true;
```
Comment (cizezsy) [inline src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java]: Maybe we should set the default value to false.

File: src/main/resources/io/jenkins/plugins/coverage/CoveragePublisher/config.jelly [hunk 5]
```diff
@@ -19,6 +19,9 @@
             <div>
                 <f:checkbox field="failNoReports" title="Failed if No Reports Found"/>
             </div>
+            <div>
+                <f:checkbox field="calculateDiffForChangeRequests" title="If it is a change request build, calculate code coverage diff with a target branch build" checked="true"/>
```
Comment (cizezsy) [inline src/main/resources/io/jenkins/plugins/coverage/CoveragePublisher/config.jelly]: Same as the above.

File: src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java (hunk 4 above)
Comment (tempora-mutantur) [reply src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java]: Agree, thanks

File: src/main/resources/io/jenkins/plugins/coverage/CoveragePublisher/config.jelly (hunk 5 above)
Comment (tempora-mutantur) [reply src/main/resources/io/jenkins/plugins/coverage/CoveragePublisher/config.jelly]: Done

**Comment by cizezsy** [comment]
@tempora-mutantur Thanks for your contribution! merged it~

**Pull request merged**